File: SortColors-75.py
# Time Complexity : O(n)
# Space Complexity : O(1)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No


# Your code here along with comments explaining your approach 

def sortColors(nums):
       
        if len(nums) == 0 or nums == None:
            return 
        
        left = 0
        right = len(nums) - 1
        
        curr = 0
        while curr <= right:
            if nums[curr] == 0:
                nums[curr], nums[left] = nums[left], nums[curr]
                left += 1
                curr += 1
            elif nums[curr] == 1:
                curr += 1
            else:
                nums[curr], nums[right] = nums[right], nums[curr]
                right -= 1
        return nums


print sortColors([2,0,2,1,1,0])

File: SortColors_75.java
// Time Complexity :  O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this :


// Your code here along with comments explaining your approach

class SortColors_75 {
    public void sortColors(int[] nums) {
        if( nums == null || nums.length == 0 ){
            return;
        }
        
        int left = 0;
        int right = nums.length-1;
        int curr = 0;
        
        while(curr <= right){
            
            if(nums[curr] == 0){
                swap(nums,curr,left);
                curr++;
                left++;
            } else if(nums[curr] == 1){
                curr++;
            } else {
                swap(nums, curr, right);
                right--;
            }
            
        }
        
    }
    
    public void swap(int[] nums, int i, int j){
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }
}

// TC: O(n)
// SC: O(1)